### ISSUE TITLE
Unhandled error event on incorrect client password

### ISSUE THREAD

**Issue by atymic** [apifytech/proxy-chain#53]
Fresh install in a clean project with a copy/paste of the exact demo script from the readme,. Happens with the `upstreamProxyUrl` set to a valid proxy or null.

Executing the following curl command: 
`curl -v -x http://bob:TopSecret@127.0.0.1:8000 https://ifconfig.co/json`
works fine. The proxy responds with the correct response (200 + data).

When I try sending an invalid password, the client receives the correct response (407). The issue is that the server crashes when this occurs (see logs for the node process below).

```
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
* Establish HTTP proxy tunnel to ifconfig.co:443
* Proxy auth using Basic with user 'bob'
> CONNECT ifconfig.co:443 HTTP/1.1
> Host: ifconfig.co:443
> Proxy-Authorization: Basic Ym9iOlRvcHNlY3JldA==
> User-Agent: curl/7.54.0
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 407 Proxy Authentication Required
< Content-Type: text/html; charset=utf-8
< Proxy-Authenticate: Basic realm="ProxyChain"
* Authentication problem. Ignoring this.
< Server: ProxyChain
< Connection: close
< Content-Length: 27
<
* Received HTTP code 407 from proxy after CONNECT
* Closing connection 0
curl: (56) Received HTTP code 407 from proxy after CONNECT
```

Logs from node process:
```
ProxyServer[8000]: Listening...
Proxy server is listening on port 8000
ProxyServer[8000]: 1 | !!! Handling CONNECT ifconfig.co:443 HTTP/1.1
ProxyServer[8000]: 1 | Using HandlerTunnelDirect
ProxyServer[8000]: 1 | Connecting to target ifconfig.co:443
ProxyServer[8000]: 1 | Target socket assigned
ProxyServer[8000]: 1 | Connected
ProxyServer[8000]: 1 | Source socket ended
ProxyServer[8000]: 1 | Closing handler
Connection 1 closed
{ srcTxBytes: 4684,
  srcRxBytes: 713,
  trgTxBytes: 550,
  trgRxBytes: 4589 }
ProxyServer[8000]: 1 | !!! Closed and removed from server
ProxyServer[8000]: 2 | !!! Handling CONNECT ifconfig.co:443 HTTP/1.1
ProxyServer[8000]: Request failed (status 407): Proxy credentials required.
events.js:174
      throw er; // Unhandled 'error' event
      ^

Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:111:27)
Emitted 'error' event at:
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)

Process finished with exit code 1
```

**Comment by atymic** [comment]
Tested on node 10 & 12, same issue.

**Comment by atymic** [comment]
I've managed a bit of a ghetto fix by adding an error handler on the socket before sending the 407 response. 

```
socket.on('error', (error) => {
	console.log(error)
	socket.end();
	socket.destroy();
});
```

I added it just about this line:
https://github.com/apifytech/proxy-chain/blob/master/src/server.js#L408

It seems to work ok, but i'm not sure if there's a socket/memory leak in there.
Either way better to fix the actual issue.

**Comment by jancurn** [comment]
Many thanks for reporting this, we'll look into it.

**Comment by atymic** [comment]
Thanks!

Let me know if you need any more info :)

**Comment by MexXxo** [comment]
I think this one is the same. If the upstreamProxy receives wrong credentials it fails with logs but no event for the implementor. Or is there anyway to handle this?

Logs:
ProxyServer[3128]: 29 | !!! Closed and removed from server
ProxyServer[3128]: 22 | Connected to upstream proxy
ProxyServer[3128]: 22 | RequestError: Invalid credentials provided for the upstream proxy., responding with custom status code 502 to client
ProxyServer[3128]: 22 | Source response finished, ending source socket
ProxyServer[3128]: 26 | Connected to upstream proxy
ProxyServer[3128]: 26 | RequestError: Invalid credentials provided for the upstream proxy., responding with custom status code 502 to client
ProxyServer[3128]: 26 | Source response finished, ending source socket
ProxyServer[3128]: 28 | Connected to upstream proxy
ProxyServer[3128]: 28 | RequestError: Invalid credentials provided for the upstream proxy., responding with custom status code 502 to client
ProxyServer[3128]: 28 | Source response finished, ending source socket
ProxyServer[3128]: 22 | Source socket finished
ProxyServer[3128]: 22 | Closing handler

**Comment by jancurn** [comment]
Sorry guys, we didn't manage to get to fixing this. PRs are more than welcome! This should be a simple fix, the unit test might be slightly more complicated but I'd say manageable.

**Comment by jancurn** [comment]
Another instance of unhandled errors observed in our production system - might be related.

```
Feb 04 00:38:08 ip-172-31-62-20 proxy@prod.log Emitted 'error' event on Socket instance at:
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  errno: 'ECONNRESET',
  code: 'ECONNRESET',
  syscall: 'read'
}
Feb 04 00:38:08 ip-172-31-62-20 proxy@prod.log [Error] Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:201:27)
Feb 04 00:38:08 ip-172-31-62-20 proxy@prod.log events.js:187
      throw er; // Unhandled 'error' event
      ^
```

**Comment by jancurn** [comment]
This bug should be fixed in the latest release `0.4.0`.
